### PR TITLE
Large overhaul

### PR DESCRIPTION
**Pull request by dcramer**
I can break this up, but I wanted to push it up for ideas before spending time.

- a Dockerfile (we should not use my namespace for it)
- The 'sync' command, and you now have to choose the subcommand
- Refactors the component into a traditional python module
- a testing framework with a simple test
- Various code refactor including normalization (black), logging, exit code cleanup, f strings, etc
- Retries on download failures (3 right now)
- direnv+poetry based dev environment
- removed channel option (appeared unused)
- visual studio code defaults

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+RUN curl -sSL https://raw.githubusercontent.com/sdispater/poetry/master/get-poetry.py | python \
+    && poetry config virtualenvs.create false
+
+COPY pyproject.toml poetry.lock /install/
+# HACK(dcramer): we need the module to be installable at this stage
+RUN mkdir protect_archiver && touch protect_archiver/__init__.py
+RUN poetry install --no-interaction --no-ansi
+
+# TODO:
```
Comment (dcramer) [inline Dockerfile]: was trying to make the docker image small, but this is as far as I got before getting frustrated

File: protect_archiver/cli/download.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+    required=False,
+    help='Start time in dateutil.parser compatible format, for example "YYYY-MM-DD HH:MM:SS+0000"',
+)
+@click.option(
+    "--end",
+    type=str,
+    required=False,
+    help='End time in dateutil.parser compatible format, for example "YYYY-MM-DD HH:MM:SS+0000"',
+)
+@click.option(
```
Comment (dcramer) [inline protect_archiver/cli/download.py]: we could move snapshot into its own command to separate logic even more at this point

File: protect_archiver/cli/sync.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        port=port,
+        username=username,
+        password=password,
+        verify_ssl=verify_ssl,
+        destination_path=dest,
+        use_subfolders=True,
+    )
+
+    # get camera list
+    print("Getting camera list")
```
Comment (dcramer) [inline protect_archiver/cli/sync.py]: note: should switch to click.echo here

File: protect_archiver/client.py [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+        self.skip_existing_files = skip_existing_files
+        self.touch_files = touch_files
+
+        self.destination_path = path.abspath(destination_path)
+
+        self.files_downloaded = 0
+        self.files_skipped = 0
+        self.max_retries = 3
+
+        self.max_downloads_with_auth = max_downloads_with_auth
```
Comment (dcramer) [inline protect_archiver/client.py]: it'd be nice to document why you might need to change these, and if we dont think you do we could just make them constants

**Comment by dcramer** [comment]
Added download speed and switched to streaming + file cleanup.

One issue I still see is its not yet offsetting fully as expected:

Downloading video for time range 2020-01-09 01:49:34.739000 - 2020-01-09 02:00:00 to /tmp/cameras/2020/01/09/Garage_1419/Garage_1419_2020-01-09--01-49-34_2020-01-09--02-00-00.mp4

**Comment by dcramer** [comment]
With sync there's a few more things I'm thinking:

1) A re-check option - I want to explore APIs a bit to see if we can do this more cheaply, but my goal is only full video chunks here for long term storage
2) The ability to continue with failures - need to determine the actual cause of 500s and similar. Right now it only continues to the next camera
3) Option to traverse from newest to oldest

**Comment by dcramer** [comment]
Alright one of the 500s was camera offline, so now itll just look for that

**Comment by dcramer** [comment]
Fixes GH-5

File: protect_archiver/cli/download.py (hunk 2 above)
Comment (danielfernau) [reply protect_archiver/cli/download.py]: agreed since it's not directly part of the download/archive process

File: pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,30 @@
+[tool.poetry]
+name = "protect-archiver"
+version = "0.1.0"
+description = ""
+authors = ["Daniel Fernau", "David Cramer"]
+license = "GPLv3"
```
Comment (danielfernau) [inline pyproject.toml]: license can be changed to "MIT" now

File: pyproject.toml [hunk 6]
```diff
@@ -0,0 +1,30 @@
+[tool.poetry]
+name = "protect-archiver"
+version = "0.1.0"
```
Comment (danielfernau) [inline pyproject.toml]: version can be changed to "2.0.0" as this is our new target

**Comment by dcramer** [comment]
For anyone paying attention, we're going to break the changes up into a few (slightly) smaller patches.

#10 is the first one.

**Pull request closed without merging**